cv_next/hooks/useUser.ts

```diff
@@ -0,0 +1,19 @@
+import { useSupabase } from "./supabase";
+import useSWR from "swr";
+
+
+export const useUser = () => {
+    const supabase = useSupabase();
+    const { data, mutate, error } = useSWR('user', supabase.auth.getUser);
+
+    const loading = !data && !error;
+    const loginState = data && !error;
```

Comment:
The combination of SSR and auth helpers messes this up... loginState is always undefined which means showing signout for signed out users

Reply:
Say, why won't it work here but will work e.g. at https://github.com/magshimim-next/cv-next/blob/0b08f2b5ed1a7a4daa5193bd62e78176770605ab/cv_next/app/cv/%5BcvId%5D/components/commentSection/commentsSection.tsx#L19
(comment section's getUser effect)?

Reply:
No entirely sure, might be due to the await? getUser gets the response from supabase so when I print data I see `TypeError: Cannot read properties of undefined (reading '_getUser')
`

Reply:
OK solved a few quirks here, note:
1. `useUser` hook now gives user data as well as use session
2. the hook uses a server action that encapsulates `supabase.auth.getUser`
3. For an example (there might be more places that uses it), I've replaced existing code that controlled the user state locally with the new hook
4. Please see that the swr doesn't work exactly as I thought - it automatically revalidates many times (by default, on every mount of the component that uses it). I've disabled most of the default automatic revalidations that swr provides, and the way to revalidate it is to manually use the `mutateUser` exposed from the hook.

Reply:
Actually now that I think about it, for client-side usage `supabase.auth.getSession` should be enough, no? The problem is with server-side right?
So I can replace the `supabase.auth.getUser` call (again encapsulated as server action) with `supabase.auth.getSession` and keep the logic in the browser without the need to do another server-client hop.
Do you agree? (have we talked about this already? exams mess with my memory)

Reply:
The approach itself is really good, will be useful as I think this also fixes #97 as the data is already managed using swr
Client-side calls to getSession are enough if we don't use it for any sensitive data as it can be tampered with quite easily. From the docs
`getUser Should always be used when checking for user authorization on the server. On the client, you can instead use getSession().session.user for faster results. getSession is insecure on the server.`
So if we just use it to get the name and photo that's fine, but we can't use that as user id in commenting etc

Reply:
So essentially there are two distinct actions here we need to differentiate:
1. data retrieval: user data, cvs, comments etc
2. data manipulation: comment post/edit, cv post, user data edit etc
The first kind can be handled in client-side supabase access, while the latter needs to be handled on the server with user validation (using `auth.getUser`). This can (should?) also be what differentiate between server actions and route handlers/client fetch.
Anyway, for now I'll simply replace the logic I've done for the user session with `getSession`, and we'll think later on how to optimize the other client/server sb separation.

Reply:
Actually nevermind let's keep it like that, their warnings are too scary :)

Reply:
Yeah keeping it like that is fine
And yes the second is important that is managed on the server using getUser as you can see from #107 #108 that bypasses this
This also fixed the other issue which is great